### ISSUE TITLE
give AVPs an option to enable inheritance to child objects

### ISSUE THREAD

**Issue by kruisdraad** [SpriteLink/NIPAP#855]
In some cases we need the AVPs to set only once in a prefix or pool and if inheritance is tagged it should propagate onto any child objects. I am thinking about a optional tag to enable this.

For example, adding a 192.168.0.0/16 with a AVPs 'AbuseMail' you want to make sure the 192.168.4.0/24 has the same AVP's unless a specific AVPs under the same name exists to override (more specific)

This would save to add the AVP's to each object if needed.

**Comment by plajjan** [comment]
I'm very hesitant to this, mainly because I don't like inheritance. I did implement it for tags, but that was somewhat arbitrary (like why not for other attributes?).

One of the downsides of inheritance is that when you change a prefix that has a lot of children, the update becomes massive (and super slow) which isn't a good thing. I've seen this in production environments and.. it's just not pretty. We could perhaps optimise that query though.

If we go back and look at the problem(s) we are trying to solve. One of them is that you might want to search across information that pertains to multiple prefixes. This is the main reason we have inherited tags, so you can search for a tag that is not directly associated with the prefix you want to find. Like you have 10.0.0.0/8 with tag "infrastructure" and then you have 10.0.1.2/31 which is a link network between foo-core-1 and bar-core-2. You now search for "#infrastructure and foo-core-1" and expect to find 10.0.1.2/31, and you will since we have inherited tags!

An alternative to solve this issue would be to allow searches to match on information in parent prefixes. Let's say you search for "parent.tag = 'infrastructure' and node='foo-core-1'" you would get prefixes where the node matches 'foo-core-1' and one of the parents (like direct parent, grand-parent, grand-grand-parent and so on) has the tag 'infrastructure'. This would allow you to do a powerful search, that solves this problem. Note how you can't do such a search today, but we could implement that.

I think we discussed this on IRC.. so for your case, where you want to find the abuse contact for a given IP address, like 192.168.4.1, the easiest thing is probably that you search for '192.168.4.1 and abusemail != ""', i.e. the abusemail attribute is not empty and the result prefixes should contain 192.168.4.1. Just grab the most specific one and it will be the one with your abuse info!

There could of course be the case when someone has failed to set abuse info. Say you have:
```
Prefix                      description                  abuseinfo
192.168.0.0/16       customer block           abuse@isp.com
    192.168.3.0/24   customer Google        abuse@google.com
    192.168.4.0/24   customer foobar         
```
So you are lucky enough to have Google as your customer and you have correctly registered 192.168.3.0/24 with the google abuse contact. For 192.168.0.0/16, which is the IP aggregate you use for all your customers, you have registered abuse@isp.com, which is your own abuse address, as a form of fallback address. Now on 192.168.4.0/24 someone forgot to fill in the abuse info, so for the search on 192.168.4.1 and with abusemail set, you would actually get back 192.168.0.0/16... I think this similar to what you are doing today, just that you are iterating over the result, and with this you could always look at the most specific match.

**Comment by kruisdraad** [comment]
Yes, currently i am doing a query for the most specific object and then ask all the parents of it as well and iterate over each object to find the format of fields needed to link abuse data. Having a more powerful search would limit the searching on large systems, but we'll need to test it either way.

And indeed, we talked about this on IRC and this would be a 'nice to have' but not essential because of the other query method ... depending if that survives large amounts of requests in very short time as a benchmark.

**Comment by plajjan** [comment]
Also, I'd like to point out that doing inheritance for tags is different compared to other attributes. First off there are two columns in the database, we have 'tags' and 'inherited_tags'. Inherited tags is always additive, meaning that it consists of the parent prefix tags + inherited_tags attributes. We never override it, we just add more tags as we progress down through the prefix structure.

For AVPs (Attribute-Value Pair, which is the internal name for "extra attributes) we don't want an additive behaviour, we want to override an inherited value with one from a prefix with a longer prefix-length, if it is filled in, otherwise we propagate the inherited value, so in essence it is avp.X OR inherited_avp.X
